### PR TITLE
Implement the lexical scoping

### PR DESCRIPTION
**Pull request by rerobika**
Signed-off-by: Robert Fancsik frobert@inf.u-szeged.hu

### CODE REVIEW

**Comment by ksh8281** [comment]
We think this is good approach for implement let, const.
How about change PR state from draft to normal If you prefer?

**Comment by clover2123** [comment]
@rerobika This patch seems good to me. Would you please change the state to `Ready for reveiw"? After then, we will work on scoping features based on this pr. Thanks:)

File: src/interpreter/ByteCodeInterpreter.cpp [hunk 1]
```diff
@@ -1343,6 +1343,39 @@ Value ByteCodeInterpreter::interpret(ExecutionState& state, ByteCodeBlock* byteC
...
+            DEFINE_OPCODE(SetConstBinding)
+                :
+            {
+                SetConstBinding* code = (SetConstBinding*)programCounter;
+                state.lexicalEnvironment()->record()->asDeclarativeEnvironmentRecord()->asDeclarativeEnvironmentRecordNotIndexed()->setAsImmutableBinding(state, code->m_name);
+                ADD_PROGRAM_COUNTER(SetConstBinding);
+                NEXT_INSTRUCTION();
+            }
+
+            DEFINE_OPCODE(NoOperation)
```
Comment (zherczeg) [inline src/interpreter/ByteCodeInterpreter.cpp]: Could we optimize this out at some point?

File: test/es2015/assert.js [hunk 2]
```diff
@@ -19,14 +19,19 @@ function assert(expression) {
...
+function assertThrows(code, errorType) {
     try {
         if (typeof code === 'function') {
             code();
         } else {
             eval(code);
         }
     } catch (e) {
+        if (errorType !== undefined) {
+            if (!(e instanceof errorType)) {
```
Comment (zherczeg) [inline test/es2015/assert.js]: No need nested if.
`if (errorType && !(e instanceof errorType)) {`

Comment (rerobika) [reply test/es2015/assert.js]: Good idea!

File: src/interpreter/ByteCodeInterpreter.cpp (hunk 1 above)
Comment (rerobika) [reply src/interpreter/ByteCodeInterpreter.cpp]: Sure, I updated the PR with removing it.

**Comment by rerobika** [comment]
@ksh8281 @clover2123 Thanks for your patience the patch is ready for the review. I think this PR can be a good base for further optimizations.

**Pull request merged**